### ISSUE TITLE
Possible problem with G-Suite Google Calendar recurring events

### ISSUE THREAD

**Issue by dstearnsadobe** [mifi/ical-expander#8]
I'm using ical-expander to parse and expand recurring events returned various calendar servers, and I noticed a very strange behavior with iCal data returned from G-Suite calendars with recurring events that I'm thinking might be a bug either in ical-expander, or possibly ical.js.

Everything works fine with my personal Google Calendar: a calendar with one recurring event returns the following iCal data, which ical-expander parses and expands as expected:

```
BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:Test
X-WR-TIMEZONE:America/Los_Angeles
X-WR-CALDESC:Test Public Calendar
BEGIN:VTIMEZONE
TZID:America/Los_Angeles
X-LIC-LOCATION:America/Los_Angeles
BEGIN:DAYLIGHT
TZOFFSETFROM:-0800
TZOFFSETTO:-0700
TZNAME:PDT
DTSTART:19700308T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:-0700
TZOFFSETTO:-0800
TZNAME:PST
DTSTART:19701101T020000
RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
DTSTART;TZID=America/Los_Angeles:20180914T140000
DTEND;TZID=America/Los_Angeles:20180914T150000
RRULE:FREQ=WEEKLY;COUNT=5;BYDAY=FR
DTSTAMP:20180911T175958Z
UID:5kg6i9kku32b0uvj1jkig3i773@google.com
CREATED:20180911T155050Z
DESCRIPTION:
LAST-MODIFIED:20180911T155050Z
LOCATION:
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Recurring Event with Ending
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART;TZID=America/Los_Angeles:20180914T100000
DTEND;TZID=America/Los_Angeles:20180914T110000
RRULE:FREQ=WEEKLY;BYDAY=FR
DTSTAMP:20180911T175958Z
UID:0240jd9trkafkccbhpl7s3t83m@google.com
CREATED:20180911T154655Z
DESCRIPTION:This is a description
LAST-MODIFIED:20180911T154656Z
LOCATION:Test Location
SEQUENCE:0
STATUS:CONFIRMED
SUMMARY:Test Recurring Event
TRANSP:OPAQUE
END:VEVENT
END:VCALENDAR
```

But when I try the same thing on a G-Suite calendar (https://calendar.google.com/calendar/ical/601t.com_a029mt9oo3j20r5okc98g4lfe4%40group.calendar.google.com/public/basic.ics), the calendar server seems to auto-expand the recurring event like this:

```
BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:Test
X-WR-TIMEZONE:America/Los_Angeles
X-WR-CALDESC:Test public calendar
BEGIN:VEVENT
DTSTART:20181012T190000Z
DTEND:20181012T200000Z
DTSTAMP:20180910T215427Z
UID:22q6ni7gkhpg31sbt2024i26og@google.com
ATTENDEE;X-NUM-GUESTS=0:mailto:601t.com_a029mt9oo3j20r5okc98g4lfe4@group.ca
 lendar.google.com
RECURRENCE-ID:20181012T190000Z
SUMMARY:Busy
END:VEVENT
BEGIN:VEVENT
DTSTART:20181005T190000Z
DTEND:20181005T200000Z
DTSTAMP:20180910T215427Z
UID:22q6ni7gkhpg31sbt2024i26og@google.com
ATTENDEE;X-NUM-GUESTS=0:mailto:601t.com_a029mt9oo3j20r5okc98g4lfe4@group.ca
 lendar.google.com
RECURRENCE-ID:20181005T190000Z
SUMMARY:Busy
END:VEVENT
BEGIN:VEVENT
DTSTART:20180928T190000Z
DTEND:20180928T200000Z
DTSTAMP:20180910T215427Z
UID:22q6ni7gkhpg31sbt2024i26og@google.com
ATTENDEE;X-NUM-GUESTS=0:mailto:601t.com_a029mt9oo3j20r5okc98g4lfe4@group.ca
 lendar.google.com
RECURRENCE-ID:20180928T190000Z
SUMMARY:Busy
END:VEVENT
BEGIN:VEVENT
DTSTART:20180921T190000Z
DTEND:20180921T200000Z
DTSTAMP:20180910T215427Z
UID:22q6ni7gkhpg31sbt2024i26og@google.com
ATTENDEE;X-NUM-GUESTS=0:mailto:601t.com_a029mt9oo3j20r5okc98g4lfe4@group.ca
 lendar.google.com
RECURRENCE-ID:20180921T190000Z
SUMMARY:Busy
END:VEVENT
BEGIN:VEVENT
DTSTART:20180914T190000Z
DTEND:20180914T200000Z
DTSTAMP:20180910T215427Z
UID:22q6ni7gkhpg31sbt2024i26og@google.com
ATTENDEE;X-NUM-GUESTS=0:mailto:601t.com_a029mt9oo3j20r5okc98g4lfe4@group.ca
 lendar.google.com
RECURRENCE-ID:20180914T190000Z
SUMMARY:Busy
END:VEVENT

... more events ...

END:VCALENDAR
```

Note how each `VEVENT` has a `RECURRENCE-ID` property that refers back to the start time of the original recurring event, but they don't have an `RRULE` nor an `EXDATE` property.

When I parse/expand the latter example using ical-expander, I get totally empty arrays for both `events` and `occurrences`.

I stepped through the code, and the problem happens at this line: 

https://github.com/mifi/ical-expander/blob/daed587f5f9d1f21110ba0d136e63aab5e5660f6/index.js#L63

The `.filter()` predicate calls the `.isRecurrenceException()` method of the ical Event object, which is here:

https://github.com/mozilla-comm/ical.js/blob/5ae4372f1fad2ef19f315e358b1f46fe4ca1ec5e/lib/ical/event.js#L301

This method returns `true` if the event has a `RECURRENCE-ID` property, which all of the events returned from the G-Suite calendar have. Thus, I get back empty arrays.

I'm not an expert on RFC5545, but I was surprised that ical.js uses just the presence of `RECURRENCE-ID` to determine that an event is a recurrence _exception_. From my reading of the spec, the `RECURRENCE-ID` is just a reference back to the original event in a recurring series, and the `EXDATE` property is used for exceptions. But perhaps this behavior is due to the way some servers tended to use `RECURRENCE-ID` in practice to identify recurrence exceptions?

Bottom line: is this a problem in ical-expander, ical.js, or the iCal data being returned from the G-Suite calendar? Thanks!

**Comment by mifi** [comment]
So How did you make these events? Like you say these seem to be auto expanded, and if you see RECURRENCE-ID for each of them points back to themselves. So they are recursions of themselves (which i don't even know if is valid in the spec), and they are all single recursions (meaning no actual recursion). When importing the ics to iCalendar, they all also show up as single recursive events. When i try to edit one of them it asks if i want to update all, and if i do, it creates a duplicate of the event on the same day. So something is spaced out there as well. I have a feeling this ical is a bit funky...

**Comment by dstearnsadobe** [comment]
As I noted, these events were returned from the following URL, which I got from a Google G-Suite calendar:

https://calendar.google.com/calendar/ical/601t.com_a029mt9oo3j20r5okc98g4lfe4%40group.calendar.google.com/public/basic.ics

So Google's G-Suite Calendar server is returning this iCal data. It looks strange to me as well, but that's what I'm getting back.

As I noted, I think this is really a problem in ical.js since they consider any event with a `RECURRENCE-ID` to be a recurrence exception, which seems odd to me.

**Comment by mifi** [comment]
I don't remember now, why it was implemented like this. But then again, why would an event have a recurrence id that points back to itself. Is it a recurring event if it has 0 recurrences? If this is a valid then I guess we could add some code that checks if it points back to itself, and then just pass it through...

So all you did in G suite was to create a single recurring event that repeats every week, and the export it to ics?

**Comment by dstearnsadobe** [comment]
Yes, all I did was create a new calendar in my G-Suite Calendar UI, create one recurring event on that calendar, and then go into the sharing settings to make it public and grab the public .ics link. For a G-Suite calendar, you have to explicitly mark it as public and choose whether to expose just free/busy or the full details before the .ics link will work.

Note that I've tried this only on our organization's particular G-Suite calendar, so I don't know if this behavior happens on *all* G-Suite calendars. It could be just ours, but suspect it would be common given that everything is hosted by Google.

Lastly, the project I needed this for was recently cancelled, so feel free to just close this if you don't think the case is worth investigating. But you might want to look into it, as I'm sure others will run into at some point if they deal with calendar data returned from a G-Suite calendar ics link.

**Comment by mifi** [comment]
By GSuite you mean paid business google account right? Which is actually the same as a private calendar just a different subscription? I tried to do the same now with my own g suite calendar and it exports a single VEVENT in the ics file, as expected.

It sounds like a bug in gsuite, because your exported ics file doesn't make sense. I've never encountered data like that. I'll close for now 👍

**Comment by peterjuras** [comment]
Hi mifi,

thanks for this great module, it makes parsing really enjoyable! 

I'm encountering the same issue. This is the calendar data I am trying to expand, it is from a (paid) G-Suite calendar:

Do you know a way to parse the recurring (apparently expanded) events?

```
(Removed)
```

**Comment by mifi** [comment]
Parsing doesn't work? can you please clarify a bit. Maybe reduce the case down to a simple case? (this ics is very long for me to read)

**Comment by peterjuras** [comment]
Ah my bad, I noticed that recurring events show up in the `occurences` property! Thanks again for creating this library!
